stateproof/builder.go

```diff
@@ -0,0 +1,474 @@
+	if err != nil {
+		return builder{}, err
+	}
+
+	return res, nil
+}
+
+func (spw *Worker) initBuilders() {
+	spw.mu.Lock()
+	defer spw.mu.Unlock()
+
+	var roundSigs map[basics.Round][]pendingSig
+	err := spw.db.Atomic(func(ctx context.Context, tx *sql.Tx) (err error) {
+		roundSigs, err = getPendingSigs(tx)
+		return
+	})
+	if err != nil {
+		spw.log.Warnf("initBuilders: getPendingSigs: %w", err)
+		return
+	}
+
+	for rnd, sigs := range roundSigs {
+		if _, ok := spw.builders[rnd]; ok {
+			spw.log.Warnf("initBuilders: round %d already present", rnd)
+			continue
+		}
+		spw.addSigsToBuilder(sigs, rnd)
+	}
+}
+
+func (spw *Worker) addSigsToBuilder(sigs []pendingSig, rnd basics.Round) {
+	builderForRound, err := spw.makeBuilderForRound(rnd)
+	if err != nil {
+		spw.log.Warnf("addSigsToBuilder: makeBuilderForRound(%d): %v", rnd, err)
+		return
+	}
+	spw.builders[rnd] = builderForRound
+
+	for _, sig := range sigs {
+		pos, ok := builderForRound.voters.AddrToPos[sig.signer]
+		if !ok {
+			spw.log.Warnf("addSigsToBuilder: cannot find %v in round %d", sig.signer, rnd)
+			continue
+		}
+
+		isPresent, err := builderForRound.Present(pos)
+		if err != nil {
+			spw.log.Warnf("addSigsToBuilder: failed to invoke builderForRound.Present on pos %d - %w ", pos, err)
+			continue
+		}
+		if isPresent {
+			spw.log.Warnf("addSigsToBuilder: cannot add %v in round %d: position %d already added", sig.signer, rnd, pos)
+			continue
+		}
+
+		if err := builderForRound.IsValid(pos, &sig.sig, false); err != nil {
+			spw.log.Warnf("addSigsToBuilder: cannot add %v in round %d: %v", sig.signer, rnd, err)
+			continue
+		}
+		if err := builderForRound.Add(pos, sig.sig); err != nil {
+			spw.log.Warnf("addSigsToBuilder: error while adding sig. inner error: %w", err)
+			continue
+		}
+	}
+}
+
+func (spw *Worker) handleSigMessage(msg network.IncomingMessage) network.OutgoingMessage {
+	var ssig sigFromAddr
+	err := protocol.Decode(msg.Data, &ssig)
+	if err != nil {
+		spw.log.Warnf("spw.handleSigMessage(): decode: %v", err)
+		return network.OutgoingMessage{Action: network.Disconnect}
+	}
+
+	fwd, err := spw.handleSig(ssig, msg.Sender)
+	if err != nil {
+		spw.log.Warnf("spw.handleSigMessage(): %v", err)
+	}
+
+	return network.OutgoingMessage{Action: fwd}
+}
+
+// handleSig adds a signature to the pending in-memory state proof provers (builders). This function is
+// also responsible for making sure that the signature is valid, and not duplicated.
+// if a signature passes all verification it is written into the database.
+func (spw *Worker) handleSig(sfa sigFromAddr, sender network.Peer) (network.ForwardingPolicy, error) {
+	spw.mu.Lock()
+	defer spw.mu.Unlock()
+
+	builderForRound, ok := spw.builders[sfa.Round]
+	if !ok {
+		latest := spw.ledger.Latest()
+		latestHdr, err := spw.ledger.BlockHdr(latest)
+		if err != nil {
+			// The latest block in the ledger should never disappear, so this should never happen
+			return network.Disconnect, err
```

Comment:
why disconnect from the peer if you have a local error loading ledger.BlockHdr(latest)? This has nothing to do with the message that was sent.

Reply:
```suggestion
			return network.Ignore, err
```

Reply:
this looks like the original compact cert code, but I agree, on internal errors (not caused by input data from a peer) we'd emit ignore action, not disconnect

Reply:
This is disconnect because the error is not just an "internal error", it is an unexpected one, and that is why we disconnect.

Reply:
but is there anything the peer did, via the message they sent, to cause this error? we have had bursts of disconnections due to bugs that caused disconnect before .. would be good to avoid..

Reply:
Can you tell why this error may happen? And then decide, if it should continue maintaining the connection.

Reply:
There are lots of errors that are handled inside algod and result in an error being logged, or the node crashing, or the task the node was trying to do being aborted. My understanding of the rule we want to enforce is: only the errors that are directly caused by a malicious network participant are supposed to result in disconnections.

Reply:
This is because, even in of one subsystem has a failure, there are still other network message handlers (such as relaying transactions and votes) that we want to continue to operate, if possible, and not be impacted to another subsystem disconnecting peers as error-handling behavior.